### PR TITLE
bpo-34939: Allow annotated global names in module namespace

### PR DESCRIPTION
**Pull request by pablogsal**
https://bugs.python.org/issue34939

### CODE REVIEW

**Comment by tirkarthi** [comment]
@ilevkivskyi Is there any [documentation](https://docs.python.org/3.8/reference/simple_stmts.html#the-global-statement) update needed with respect to https://bugs.python.org/issue34939#msg327386 ?

Thanks

File: Python/symtable.c [hunk 1]
```diff
@@ -1174,8 +1174,9 @@ symtable_visit_stmt(struct symtable *st, stmt_ty s)
             if (cur < 0) {
                 VISIT_QUIT(st, 0);
             }
-            if ((cur & (DEF_GLOBAL | DEF_NONLOCAL))
-                && s->v.AnnAssign.simple) {
+            if (((cur & DEF_NONLOCAL) |
+                 ((cur & DEF_GLOBAL) && !(cur & DEF_GLOBAL_TOP))
+                ) && s->v.AnnAssign.simple) {
```
Comment (ilevkivskyi) [inline Python/symtable.c]: Adding a new symbol table flag to solve this looks like an overkill (plus TBH I don't fully understand this solution). Why not just skip emitting this error at global scope? For example check that `st->st_cur->ste_symbols != st->st_global`?

File: Lib/test/test_grammar.py [hunk 2]
```diff
@@ -301,6 +302,12 @@ def one():
         [list][0]: type
         my_lst[one()-1]: int = 5
         self.assertEqual(my_lst, [5])
+        exec(textwrap.dedent("""
+            def f():
+                global x
+                x = 1
+            x:int = 3
+            """))
```
Comment (ilevkivskyi) [inline Lib/test/test_grammar.py]: I would reserve this test file only for basic things. I propose to add this test to either `test_symtable.py` or `test_syntax.py`.

Also there is even a simpler repro for the issue:
```python
global x
x: int  # SyntaxError
```
plus I would add test cases for `nonlocal` as well (even though it already works).

**Review by ilevkivskyi — COMMENTED** [review]
Thanks for working on this! Here are some comments.

File: Python/symtable.c (hunk 1 above)
Comment (ilevkivskyi) [reply Python/symtable.c]: I would probably also add a comment here explaining that annotation forces a variable to be local for current scope, while `global` forces it to be global, so these two "directives" are conflicting, _unless_ we are currently at the global scope.

Comment (pablogsal) [reply Python/symtable.c]: Thanks for the review @ilevkivskyi. This is the first time I touch the symtable so sorry if the original proposed solution was a bit convoluted. I have simplified it in 701bffe.

File: Lib/test/test_grammar.py (hunk 2 above)
Comment (pablogsal) [reply Lib/test/test_grammar.py]: It seems that `test_syntax` tests all raise SyntaxErrors, but in this case, we want to test that the solution is valid. On the other hand, `test_symtable` seems to test the symtable properties directly, as opposed to check that something is a valid Syntax or not. I am a bit unsure on which file to put the reproducer (along with the "nonlocal" one).

Comment (pablogsal) [reply Lib/test/test_grammar.py]: I have moved the tests to `test_symtable` in 0e21eba.

File: Lib/test/test_symtable.py [hunk 3]
```diff
@@ -144,6 +144,16 @@ def test_annotated(self):
...
 
+        # Test that annotations in the global scope are valid after the
+        # variable is declared as nonlocal.
+        st5 = symtable.symtable('global x\nx: int', 'test', 'exec')
+        self.assertTrue(st5.lookup("x").is_global())
+
+        # Test that annotations for nonlocals are valid after the
+        # variable is declared as nonlocal.
+        st6 = symtable.symtable('def g():\n    x=2\n    def f():\n'
+                                '        nonlocal x\n    x:int', 'test', 'exec')
```
Comment (ilevkivskyi) [inline Lib/test/test_symtable.py]: Complex multi-line tests are easier to read if you split them like this:
```python
        st6 = symtable.symtable('def g():\n'
                                '    x = 2\n'
                                '    def f():\n'
                                '        nonlocal x\n'
                                '    x: int',
                                'test', 'exec')
```
(Also I would stick to PEP 8 even in tests, see other test above. In this case space around `=` and space after `:` in annotation.)

File: Lib/test/test_grammar.py (hunk 2 above)
Comment (pablogsal) [reply Lib/test/test_grammar.py]: I have openend https://github.com/python/cpython/pull/9872 as there are no current way to test for nonlocal declarations in the symtable python module.

**Review by ilevkivskyi — APPROVED** [review]
Thanks! I have one additional suggestion. Feel free to merge after you address it.

File: Lib/test/test_symtable.py (hunk 3 above)
Comment (pablogsal) [reply Lib/test/test_symtable.py]: Done in 547dbba. I would like to add an extra nonlocal check to `st6` after if #9872 is merged.

**Comment by pablogsal** [comment]
@ilevkivskyi Thank you very much for the review and the helpful advice :)

**Pull request merged**